shopfloor/services/delivery.py

```diff
@@ -123,16 +451,78 @@ def reset_qty_done_line(self, picking_id, line_id):
         Transitions:
         * deliver: always return here with updated data
         """
-        return self._response()
+        picking = self.env["stock.picking"].browse(picking_id).exists()
+        if picking:
+            response = self._check_picking_status(picking)
+            if response:
+                return response
+        else:
+            return self._response_for_deliver(
+                message=self.msg_store.stock_picking_not_found()
+            )
+        line = self.env["stock.move.line"].browse(move_line_id).exists()
+        if line:
+            if line.picking_id != picking:
+                return self._response_for_deliver(
+                    picking=picking,
+                    message=self.msg_store.line_not_available_in_picking(picking),
+                )
+            if line.package_id:
+                return self._response_for_deliver(
+                    picking=picking,
+                    message=self.msg_store.line_has_package_scan_package(),
+                )
+            self._reset_lines(line)
+            return self._response_for_deliver(picking)
+        return self._response_for_deliver(
+            picking=picking, message=self.msg_store.record_not_found(),
+        )
 
-    def done(self, picking_id):
+    def done(self, picking_id, confirm=False):
```

Comment:
@jgrandguillaume we said we would need multiple pickings here to confirm but as `_action_picking_done` it's called every time once a line is processed you'll get pickings done anyway before getting to the screen "confirm shipping".
In the end you would see only the one that were not set as done because of partial delivery.
Hence, I'm not sure the new screen will make sense.

Reply:
Well I guess we originally though this process to validate picking by picking then, making the "Mark as done" button available when showing a picking details.
Need to check with @jbaudoux

Reply:
yes but the btn is useless if you have other lines to be processed in the other WH.
And the automatic "mark as done" logic satisfy the need I think.

Reply:
After discussion @simahawk @sebalix We propose:
 - Allow to give a list of pickings
 - Start a save point
 - Process all picking

 -> If all entirely validated, fine
-> If at least one picking ask for a back order -> Ask the user a confirmation
    -> Says yes : Validate them all with back orders
    - Says no: Rolle back everything and confirm nothing

Reply:
@jgrandguillaume we should discuss this... actually what we discussed when I drafted the specs did seem good at the time, I summarize what we discussed:

* every time we scan a line or manually add it (by clicking on it), if it's the last line of the transfer, it's set to done automatically
* when a transfer is shown on the screen (because we scanned something or because we used the manual selection screen), we have a button to set it to done, which is used only for the exceptions when we have to force a partial (truck is full, ...), so doesn't have to be used a lot as all the other ones will be set to done progressively

Reply:
So we keep things as they are?

Reply:
Answer: yes